keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/sql/LegacySpecUpgradeTests.kt

```diff
@@ -0,0 +1,137 @@
+package com.netflix.spinnaker.keel.sql
+
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceKind
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.api.plugins.SupportedKind
+import com.netflix.spinnaker.keel.resources.ResourceSpecIdentifier
+import com.netflix.spinnaker.keel.resources.SpecMigrator
+import com.netflix.spinnaker.keel.serialization.configuredObjectMapper
+import com.netflix.spinnaker.keel.test.resource
+import com.netflix.spinnaker.kork.sql.config.RetryProperties
+import com.netflix.spinnaker.kork.sql.config.SqlRetryProperties
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import java.time.Clock.systemUTC
+import java.time.Instant
+import java.time.Instant.EPOCH
+import strikt.api.expectCatching
+import strikt.assertions.isA
+import strikt.assertions.isEqualTo
+import strikt.assertions.succeeded
+
+internal class LegacySpecUpgradeTests : JUnit5Minutests {
+
+  data class SpecV1(
+    val name: String
+  ) : ResourceSpec {
+    override val id = name
+    override val application = "fnord"
+  }
+
+  data class SpecV2(
+    val name: String,
+    val number: Int
+  ) : ResourceSpec {
+    override val id = name
+    override val application = "fnord"
+  }
+
+  data class SpecV3(
+    val name: String,
+    val number: Int,
+    val timestamp: Instant
+  ) : ResourceSpec {
+    override val id = name
+    override val application = "fnord"
+  }
+
+  object Fixture {
+    @JvmStatic
+    val testDatabase = initTestDatabase()
+
+    private val jooq = testDatabase.context
+    private val retryProperties = RetryProperties(1, 0)
+    private val sqlRetry = SqlRetry(SqlRetryProperties(retryProperties, retryProperties))
+
+    val v1 = SupportedKind(ResourceKind.parseKind("test/whatever@v1"), SpecV1::class.java)
+    val v2 = SupportedKind(ResourceKind.parseKind("test/whatever@v2"), SpecV2::class.java)
+    val v3 = SupportedKind(ResourceKind.parseKind("test/whatever@v3"), SpecV3::class.java)
+
+    val resourceTypeIdentifier = ResourceSpecIdentifier(v1, v2, v3)
+
+    val v1to2Migrator = object : SpecMigrator<SpecV1, SpecV2> {
+      override val input = v1
+      override val output = v2
+
+      override fun migrate(spec: SpecV1): SpecV2 =
+        SpecV2(spec.name, 1)
+    }
+    val v2to3Migrator = object : SpecMigrator<SpecV2, SpecV3> {
```

Comment:
Nice, this looks really easy to use.

Reply:
Thanks. Once I got going on it, this was easy to implement too. I was overthinking it last week. @luispollo suggested to use the actual types as well which made it much nicer.